src/Lumper.Lib/BSP/Struct/PakFileEntry.cs

```diff
@@ -7,50 +7,98 @@ namespace Lumper.Lib.BSP.Struct;
 using Newtonsoft.Json;
 using SharpCompress.Archives.Zip;
 
-public class PakFileEntry
+/// <summary>
+/// A pakfile entry derived from either a ZipArchiveEntry or just a stream.
+/// We assume one is non-null throughout.
+/// </summary>
+public sealed class PakfileEntry : IDisposable
 {
-    public PakFileEntry(ZipArchiveEntry entry)
+    public PakfileEntry(ZipArchiveEntry zipEntry)
     {
-        _entry = entry;
-        Key = entry.Key;
+        ZipEntry = zipEntry;
+        Key = zipEntry.Key ?? throw new InvalidDataException("Pakfile contains an item without a key");
     }
 
-    public PakFileEntry(string key, Stream stream)
+    public PakfileEntry(string key, Stream stream)
     {
         Key = key;
-        _dataStream = stream;
+
+        stream.Seek(0, SeekOrigin.Begin);
+        var mem = new MemoryStream();
+        stream.CopyTo(mem);
+        _buffer = mem.ToArray();
     }
 
+    // Probably shouldn't be public but whatever. be careful!
     [JsonIgnore]
-    private readonly ZipArchiveEntry? _entry;
+    public ZipArchiveEntry? ZipEntry { get; }
 
     public string Key { get; set; }
 
+    [JsonProperty]
+    public long? CompressedSize => ZipEntry?.CompressedSize ?? null;
+
     [JsonIgnore]
-    private Stream? _dataStream;
+    public bool IsModified { get; set; }
+
+    [JsonIgnore]
+    private byte[]? _buffer;
+
+    private readonly List<MemoryStream> _issuedStreams = [];
 
     [JsonIgnore]
-    public Stream DataStream
+    private readonly object _lock = new();
+
+    /// <summary>
+    /// Get an stream to the uncompressed pakfile entry.
+    ///
+    /// This stream can be invalidated if the entry is updated by the user.
+    /// </summary>
+    public MemoryStream GetReadOnlyStream()
+    {
+        lock (_lock)
+        {
+            if (_buffer is not null)
+                return IssueStream();
+
+            using Stream zipStream = ZipEntry!.OpenEntryStream();
+            using var outStream = new MemoryStream();
+            zipStream.CopyTo(outStream);
+            _buffer = outStream.ToArray();
+            return IssueStream();
+        }
+    }
+
+    private MemoryStream IssueStream()
     {
-        get => _dataStream ?? _entry!.OpenEntryStream(); // Ctors take either entry or stream, one is non-null
-        set
+        var stream = new MemoryStream(_buffer!, writable: false);
+        _issuedStreams.Add(stream);
+        return stream;
+    }
```

Comment:
don't understand why we need a list of streams .. doesn't this just copy the same data in here and when the data gets changed all are disposed anyway?

Would also need the lock

Reply:
Because we have multiple threads reading and seeking.

I think IssueStream is a bit unnecessary for saving repeating three lines of code inside the same method, getting rid of it. Though just to clarify, IssueStream was only getting called from inside locks, isn't a lock here unnecessary? Code in question since force-push will alert this:
```cs
    public MemoryStream GetReadOnlyStream()
    {
        lock (_lock)
        {
            if (_buffer is not null)
                return IssueStream();

            using Stream zipStream = ZipEntry!.OpenEntryStream();
            using var outStream = new MemoryStream();
            zipStream.CopyTo(outStream);
            _buffer = outStream.ToArray();
            return IssueStream();
        }
    }

    private MemoryStream IssueStream()
    {
        var stream = new MemoryStream(_buffer!, writable: false);
        _issuedStreams.Add(stream);
        return stream;
    }
```

Reply:
lock unnecessary: yea I guess your right, didn't check where it was used and didn't even realize that its private .. saw the dispose code and not having a lock around the Add seemed wrong. Sorry about that.